### ISSUE TITLE
Package Deployer task fails with Timeout

### ISSUE THREAD

**Issue by hajekj** [WaelHamze/dyn365-ce-vsts-tasks#173]
Hello, we are using the `Package Deployer` task from your tools (v12) in Azure DevOps. The task is configured followingly:
- Package Deployment Timeout: 00:30:00
- Connection Timeout: 60

However after ~11 minutes into the package deployment, the task fails with timeout:

```
2020-04-19T09:02:10.6775425Z PackageDeployer Verbose: 16 : Message: SolutionName1 - UploadHoldingSolutionV2 - SkipImportJobCheck=False
2020-04-19T09:02:10.7022901Z PackageDeployer Information: 8 : Message: SolutionName1 - Import Progress is : 100
2020-04-19T09:02:10.7045009Z PackageDeployer Information: 8 : Message: SolutionName1 Solution Import Result = success - Duration: 00:02:00.4437293
2020-04-19T09:02:10.7332238Z PackageDeployer Information: 8 : Message: SolutionName1 - Updated solution Version: 1.0.2.51 to Version : 1.0.2.62
2020-04-19T09:02:10.7338453Z PackageDeployer Information: 8 : Message: SolutionName1 : User Code Execution : OP=RunSolutionUpgradeMigrationStep : Status=Execute
2020-04-19T09:02:10.7344922Z PackageDeployer Information: 8 : Message: SolutionName1 : User Code Execution : OP=RunSolutionUpgradeMigrationStep : Status=Complete : Duration=00:00:00.0006318
2020-04-19T09:02:10.7350319Z PackageDeployer Information: 8 : Message: SolutionName1 - Updating solution Version: 1.0.2.51 to Version : 1.0.2.62 - Deleting Holding Solution
2020-04-19T09:02:10.7355701Z PackageDeployer Information: 8 : Message: SolutionName1 - Executing Solution Import - PromoteHoldingSolutionV2
2020-04-19T09:04:10.8106265Z PackageDeployer Verbose: 16 : Message: solution does not exist on server - solutionName = SolutionName1 - targetEnviromentSolutionVersion = 1.0.2.51 - incomingSolutionVersion = 1.0.2.62
2020-04-19T09:04:10.8115304Z PackageDeployer Error: 2 : Message: SolutionName1 - Updating solution Version: 1.0.2.51 to Version : 1.0.2.62 - Deleting Holding Solution FAILED :  : ************ TimeoutException - DeleteAndPromote : Delete and Promote Solution to New version(DeleteAndPromoteSolutionRequest) |=> The request channel timed out while waiting for a reply after 00:02:00. Increase the timeout value passed to the call to Request or increase the SendTimeout value on the Binding. The time allotted to this operation may have been a portion of a longer timeout.
2020-04-19T09:04:10.8118230Z The request channel timed out while waiting for a reply after 00:02:00. Increase the timeout value passed to the call to Request or increase the SendTimeout value on the Binding. The time allotted to this operation may have been a portion of a longer timeout. => The HTTP request to 'https://customerxxx.crm4.dynamics.com/XRMServices/2011/Organization.svc/web?SDKClientVersion=9.0.44.892' has exceeded the allotted timeout of 00:02:00. The time allotted to this operation may have been a portion of a longer timeout. => The operation has timed out[TerminalFailure] Failed to Execute Command - DeleteAndPromote : RequestID=371f00d1-297b-4c3a-8b65-7cb2136c81dd : Delete and Promote Solution to New version(DeleteAndPromoteSolutionRequest) duration=00:02:00.0107310 ExceptionMessage = The request channel timed out while waiting for a reply after 00:02:00. Increase the timeout value passed to the call to Request or increase the SendTimeout value on the Binding. The time allotted to this operation may have been a portion of a longer timeout.
2020-04-19T09:04:10.8121116Z The request channel timed out while waiting for a reply after 00:02:00. Increase the timeout value passed to the call to Request or increase the SendTimeout value on the Binding. The time allotted to this operation may have been a portion of a longer timeout. => The HTTP request to 'https://customerxxx.crm4.dynamics.com/XRMServices/2011/Organization.svc/web?SDKClientVersion=9.0.44.892' has exceeded the allotted timeout of 00:02:00. The time allotted to this operation may have been a portion of a longer timeout. => The operation has timed out
2020-04-19T09:04:10.8122345Z Source	: mscorlib
2020-04-19T09:04:10.8122516Z Method	: HandleReturnMessage
2020-04-19T09:04:10.8122689Z Date	: 9:04:10 AM
2020-04-19T09:04:10.8122881Z Time	: 4/19/2020
2020-04-19T09:04:10.8123413Z Error	: The request channel timed out while waiting for a reply after 00:02:00. Increase the timeout value passed to the call to Request or increase the SendTimeout value on the Binding. The time allotted to this operation may have been a portion of a longer timeout.
2020-04-19T09:04:10.8123976Z Stack Trace	: Server stack trace: 
2020-04-19T09:04:10.8124303Z    at System.ServiceModel.Channels.RequestChannel.Request(Message message, TimeSpan timeout)
2020-04-19T09:04:10.8124672Z    at System.ServiceModel.Dispatcher.RequestChannelBinder.Request(Message message, TimeSpan timeout)
2020-04-19T09:04:10.8125136Z    at System.ServiceModel.Channels.ServiceChannel.Call(String action, Boolean oneway, ProxyOperationRuntime operation, Object[] ins, Object[] outs, TimeSpan timeout)
2020-04-19T09:04:10.8125781Z    at System.ServiceModel.Channels.ServiceChannelProxy.InvokeService(IMethodCallMessage methodCall, ProxyOperationRuntime operation)
2020-04-19T09:04:10.8126177Z    at System.ServiceModel.Channels.ServiceChannelProxy.Invoke(IMessage message)
2020-04-19T09:04:10.8126368Z 
2020-04-19T09:04:10.8126559Z Exception rethrown at [0]: 
2020-04-19T09:04:10.8126847Z    at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
2020-04-19T09:04:10.8127215Z    at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
2020-04-19T09:04:10.8127596Z    at Microsoft.Xrm.Sdk.IOrganizationService.Execute(OrganizationRequest request)
2020-04-19T09:04:10.8128030Z    at Microsoft.Xrm.Sdk.WebServiceClient.WebProxyClient`1.ExecuteAction[TResult](Func`1 action)
2020-04-19T09:04:10.8128432Z    at Microsoft.Xrm.Tooling.Connector.CrmServiceClient.CrmCommand_Execute(OrganizationRequest req, String errorStringCheck)
2020-04-19T09:04:10.8128901Z ======================================================================================================================
2020-04-19T09:04:10.8129208Z Inner Exception Level 1	: 
2020-04-19T09:04:10.8129392Z Source	: System.ServiceModel
2020-04-19T09:04:10.8129589Z Method	: ProcessGetResponseWebException
2020-04-19T09:04:10.8129813Z Date	: 9:04:10 AM
2020-04-19T09:04:10.8129962Z Time	: 4/19/2020
2020-04-19T09:04:10.8130490Z Error	: The HTTP request to 'https://customerxxx.crm4.dynamics.com/XRMServices/2011/Organization.svc/web?SDKClientVersion=9.0.44.892' has exceeded the allotted timeout of 00:02:00. The time allotted to this operation may have been a portion of a longer timeout.
2020-04-19T09:04:10.8131290Z Stack Trace	: at System.ServiceModel.Channels.HttpChannelUtilities.ProcessGetResponseWebException(WebException webException, HttpWebRequest request, HttpAbortReason abortReason)
2020-04-19T09:04:10.8131813Z    at System.ServiceModel.Channels.HttpChannelFactory`1.HttpRequestChannel.HttpChannelRequest.WaitForReply(TimeSpan timeout)
2020-04-19T09:04:10.8132247Z    at System.ServiceModel.Channels.RequestChannel.Request(Message message, TimeSpan timeout)
2020-04-19T09:04:10.8132760Z ======================================================================================================================
2020-04-19T09:04:10.8133077Z Inner Exception Level 2	: 
2020-04-19T09:04:10.8133249Z Source	: System
2020-04-19T09:04:10.8133445Z Method	: GetResponse
2020-04-19T09:04:10.8133605Z Date	: 9:04:10 AM
2020-04-19T09:04:10.8133759Z Time	: 4/19/2020
2020-04-19T09:04:10.8133975Z Error	: The operation has timed out
2020-04-19T09:04:10.8134217Z Stack Trace	: at System.Net.HttpWebRequest.GetResponse()
2020-04-19T09:04:10.8134587Z    at System.ServiceModel.Channels.HttpChannelFactory`1.HttpRequestChannel.HttpChannelRequest.WaitForReply(TimeSpan timeout)
2020-04-19T09:04:10.8135087Z ======================================================================================================================
2020-04-19T09:04:10.8195184Z PackageDeployer Verbose: 16 : Message: package deployment failure exception type : System.TimeoutException
2020-04-19T09:04:10.8201815Z PackageDeployer Verbose: 16 : Message: converting exception to error details
2020-04-19T09:04:10.8207938Z PackageDeployer Verbose: 16 : Message: package deployment failure exception type : System.TimeoutException
2020-04-19T09:04:10.8213506Z PackageDeployer Verbose: 16 : Message: converting exception to error details
2020-04-19T09:04:10.8218967Z PackageDeployer Verbose: 16 : Message: package deployment failure exception type : System.Net.WebException
2020-04-19T09:04:10.8224859Z PackageDeployer Verbose: 16 : Message: converting exception to error details
2020-04-19T09:04:10.8255940Z PackageDeployer Information: 8 : Message: errorDetails - ErrorCode  - -2146233083
2020-04-19T09:04:10.8262319Z PackageDeployer Information: 8 : Message: errorDetails - ErrorName  - NotProvided
2020-04-19T09:04:10.8270079Z PackageDeployer Information: 8 : Message: errorDetails - Message  - The request channel timed out while waiting for a reply after 00:02:00. Increase the timeout value passed to the call to Request or increase the SendTimeout value on the Binding. The time allotted to this operation may have been a portion of a longer timeout.
2020-04-19T09:04:10.8276207Z PackageDeployer Information: 8 : Message: errorDetails - Type  - Application
2020-04-19T09:04:10.8282688Z PackageDeployer Information: 8 : Message: errorDetails - StatusCode  - 500
2020-04-19T09:04:10.8289113Z PackageDeployer Information: 8 : Message: errorDetails - Source  - PackageDeployer
2020-04-19T09:04:10.8298286Z PackageDeployer Information: 8 : Message: errorDetails - TraceText  - ......Server stack trace: ......   at System.ServiceModel.Channels.RequestChannel.Request(Message message, TimeSpan timeout)......   at System.ServiceModel.Dispatcher.RequestChannelBinder.Request(Message message, TimeSpan timeout)......   at System.ServiceModel.Channels.ServiceChannel.Call(String action, Boolean oneway, ProxyOperationRuntime operation, Object[] ins, Object[] outs, TimeSpan timeout)......   at System.ServiceModel.Channels.ServiceChannelProxy.InvokeService(IMethodCallMessage methodCall, ProxyOperationRuntime operation)......   at System.ServiceModel.Channels.ServiceChannelProxy.Invoke(IMessage message)............Exception rethrown at [0]: ......   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)......   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)......   at Microsoft.Xrm.Sdk.IOrganizationService.Execute(OrganizationRequest request)......   at Microsoft.Xrm.Sdk.WebServiceClient.WebProxyClient`1.ExecuteAction[TResult](Func`1 action)......   at Microsoft.Xrm.Tooling.Connector.CrmServiceClient.CrmCommand_Execute(OrganizationRequest req, String errorStringCheck)
2020-04-19T09:04:10.8302447Z PackageDeployer Information: 8 : Message: errorDetails - ErrorCode  - -2146233083
2020-04-19T09:04:10.8308370Z PackageDeployer Information: 8 : Message: errorDetails - ErrorName  - NotProvided
2020-04-19T09:04:10.8314722Z PackageDeployer Information: 8 : Message: errorDetails - Message  - The HTTP request to 'https://customerxxx.crm4.dynamics.com/XRMServices/2011/Organization.svc/web?SDKClientVersion=9.0.44.892' has exceeded the allotted timeout of 00:02:00. The time allotted to this operation may have been a portion of a longer timeout.
2020-04-19T09:04:10.8320540Z PackageDeployer Information: 8 : Message: errorDetails - Type  - Application
2020-04-19T09:04:10.8326436Z PackageDeployer Information: 8 : Message: errorDetails - StatusCode  - 500
2020-04-19T09:04:10.8332278Z PackageDeployer Information: 8 : Message: errorDetails - Source  - PackageDeployer
2020-04-19T09:04:10.8338792Z PackageDeployer Information: 8 : Message: errorDetails - TraceText  -    at System.ServiceModel.Channels.HttpChannelUtilities.ProcessGetResponseWebException(WebException webException, HttpWebRequest request, HttpAbortReason abortReason)......   at System.ServiceModel.Channels.HttpChannelFactory`1.HttpRequestChannel.HttpChannelRequest.WaitForReply(TimeSpan timeout)......   at System.ServiceModel.Channels.RequestChannel.Request(Message message, TimeSpan timeout)
2020-04-19T09:04:10.8344248Z PackageDeployer Information: 8 : Message: errorDetails - ErrorCode  - -2146233079
2020-04-19T09:04:10.8350649Z PackageDeployer Information: 8 : Message: errorDetails - ErrorName  - NotProvided
2020-04-19T09:04:10.8357027Z PackageDeployer Information: 8 : Message: errorDetails - Message  - The operation has timed out
2020-04-19T09:04:10.8363094Z PackageDeployer Information: 8 : Message: errorDetails - Type  - Application
2020-04-19T09:04:10.8368718Z PackageDeployer Information: 8 : Message: errorDetails - StatusCode  - 500
2020-04-19T09:04:10.8374266Z PackageDeployer Information: 8 : Message: errorDetails - Source  - PackageDeployer
2020-04-19T09:04:10.8380472Z PackageDeployer Information: 8 : Message: errorDetails - TraceText  -    at System.Net.HttpWebRequest.GetResponse()......   at System.ServiceModel.Channels.HttpChannelFactory`1.HttpRequestChannel.HttpChannelRequest.WaitForReply(TimeSpan timeout)
2020-04-19T09:04:10.8385910Z PackageDeployer Information: 8 : Message: RaiseFailEvent - update progress with fail event
2020-04-19T09:04:10.8416004Z PackageDeployer Information: 8 : Message: ****** PACKAGE DEPLOYMENT PROCESS COMPLETED. Result:FAILED  Duration:00:11:13.0257029 ******
2020-04-19T09:04:10.8421631Z PackageDeployer Information: 8 : Message: SolutionName1 - Pausing for 5 seconds before resuming operation. Retry# 0
```

It appears that the timeout is not respected or passed to Package Deployer properly, since the deploy of the solution fails after 2 minutes of the solution deploy (this solution takes about 10 minutes to import, and about 5 to upgrade). The package deployer waits correctly for the import, but doesn't seem to wait for the upgrade to complete and cuts it after 2 minutes.

So my question is, whether this is a bug in the tasks or Microsoft's Package Deployer... From my preliminary search, I couldn't find any issue with the task itself, but I may be wrong.

**Comment by WaelHamze** [comment]
@hajekj 

Package Deployment Timeout: is passed into the Import-CrmPackage Cmdlet (-MaxCrmConnectionTimeOutMinutes)
Connection Timeout: is passed into the Get-CrmConnection Cmdlet (-Timeout)

I don't see any other timeout option in the package deployer interface.

I would suggest you try to import the package from the powershell interface directly and see if you get any issues.

**Comment by hajekj** [comment]
So I actually figured out the cause, and waiting for bug submission to Microsoft. We are using Certificate authentication against Azure AD, however, the Get-CrmConnection only sets the passed timeout to OAuth connection:

![image](https://user-images.githubusercontent.com/8337913/79731349-75e72300-82f2-11ea-8845-580d0fe3e2da.png)

So this appears to be the cause. I wish they just made the tools OSS, so I wouldn't have to spend hours in ILSpy to figure it out.

**Comment by WaelHamze** [comment]
@hajekj thanks for sharing, let me know what they tell you

**Comment by hajekj** [comment]
I have just been told that it should be fixed in next release flight. So hopefully it will be soon. Will keep this thread updated.

**Comment by hajekj** [comment]
This is now fixed with [3.3.0.890](https://www.powershellgallery.com/packages/Microsoft.Xrm.Tooling.CrmConnector.PowerShell/3.3.0.890)

**Comment by WaelHamze** [comment]
@hajekj thanks for the update. You can use a specific version of the PS Modules by configuring the Tool Installer task.
